test/HaveLetLinter.lean

```diff
@@ -0,0 +1,144 @@
+import Mathlib.Tactic.Linter.HaveLetLinter
+import Mathlib.Tactic.Tauto
+
+/-- a tactic that simply logs an empty message.  Useful for testing the chattiness of the
+`haveLet` linter. -/
+elab "noise" : tactic => do Lean.logInfo ""
+
+set_option linter.haveLet 2 in
+#guard_msgs in
+-- check that `tauto`, `replace`, `classical` are ignored
+example : True := by
+  classical
+  let zero := 0
+  replace _zero := zero
+  let eq := (rfl : 0 = 0)
+  replace _eq := eq
+  tauto
+
+/--
+info:
+
+---
+warning: '_zero : Nat' is a Type and not a Prop. Consider using 'let' instead of 'have'.
+[linter.haveLet]
+-/
+#guard_msgs in
```

Comment:
I'm a bit confused about linting level `1`:

If I just comment the `#guard_msgs in`, the linter does not highlight the `have`, and neither does it if I add `set_option linter.haveLet 1`. Only if `set_option linter.haveLet 2` is chosen the linter will display the warning.

Was your intention not that having `set_option linter.haveLet 1` together with `noise` somewhere in the proof would cause the linting warning to be displayed?

Reply:
I am also confused by this: you understood correctly what the linter is supposed to do.  However, the message that `noise` was logging was somehow invisible to the linter (I do not undestand why, probably because it is "inside" the `TacticM` and does not reach the external state).

Anyway, I replaced `noise` with a macro that simply introduces a vacuous `sorry` and now, hopefully the tests align with what I intended and also what you expected!